File: applications/rtkvarianprobeamgeometry/rtkvarianprobeamgeometry.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkvarianprobeamgeometry_ggo.h"
#include "rtkMacro.h"
#include "rtkVarianProBeamGeometryReader.h"
#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkGgoFunctions.h"

int main(int argc, char * argv[])
{
  GGO(rtkvarianprobeamgeometry, args_info);

  // Create geometry reader
  rtk::VarianProBeamGeometryReader::Pointer reader;
  reader = rtk::VarianProBeamGeometryReader::New();
  reader->SetXMLFileName(args_info.xml_file_arg);
  reader->SetProjectionsFileNames( rtk::GetProjectionsFileNamesFromGgo(args_info) );
  TRY_AND_EXIT_ON_ITK_EXCEPTION( reader->UpdateOutputData() )

  // Write
  rtk::ThreeDCircularProjectionGeometryXMLFileWriter::Pointer xmlWriter =
    rtk::ThreeDCircularProjectionGeometryXMLFileWriter::New();
  xmlWriter->SetFilename(args_info.output_arg);
  xmlWriter->SetObject( reader->GetGeometry() );
  TRY_AND_EXIT_ON_ITK_EXCEPTION( xmlWriter->WriteFile() )

  return EXIT_SUCCESS;
}

File: applications/rtkvarianprobeamgeometry/rtkvarianprobeamgeometry.py
#!/usr/bin/env python
import argparse
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description=("Creates an RTK geometry file from a Varian ProBeam acquisition.")
    )

    parser.add_argument(
        "--xml_file",
        "-x",
        help="Varian ProBeam XML information file on projections",
        required=True,
    )
    parser.add_argument(
        "--output",
        "-o",
        help="Output file name",
        required=True,
    )
    rtk.add_rtkinputprojections_group(parser)

    return parser


def process(args_info: argparse.Namespace):
    # Create geometry reader
    if args_info.verbose:
        print(f"Reading Varian ProBeam XML: {args_info.xml_file}")

    reader = rtk.VarianProBeamGeometryReader.New()
    reader.SetXMLFileName(args_info.xml_file)

    fileNames = rtk.GetProjectionsFileNamesFromArgParse(args_info)
    reader.SetProjectionsFileNames(fileNames)

    reader.UpdateOutputData()

    # Write geometry
    if args_info.verbose:
        print(f"Writing geometry to: {args_info.output}")
    rtk.write_geometry(reader.GetGeometry(), args_info.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    return process(args_info)


if __name__ == "__main__":
    main()
